clash-prelude/src/Clash/Sized/Vector.hs

```diff
@@ -1495,6 +1496,36 @@ iterateI f a = xs
     ws = map f (lazyV xs)
 {-# INLINE iterateI #-}
 
+-- | \"'unfoldr @n f s@\" builds a vector of length @n@ from a seed value @s@,
+-- where every element @a@ is created by successive calls of @f@ on @s@. Unlike
+-- 'Data.List.unfoldr' from "Data.List" the generating function @f@ cannot
+-- dictate the length of the resulting vector, it must be statically known.
+--
+-- a simple use of 'unfoldr':
+--
+-- >>> unfoldr d10 (\s -> (s,s-1)) 10
+-- <10,9,8,7,6,5,4,3,2,1>
+unfoldr :: SNat n -> (s -> (a,s)) -> s -> Vec n a
+unfoldr SNat = unfoldrI
+{-# INLINE unfoldr #-}
+
+-- | \"'unfoldr @f s@\" builds a vector from a seed value @s@, where every
+-- element @a@ is created by successive calls of @f@ on @s@; the length of the
+-- vector is inferred from the context. Unlike 'Data.List.unfoldr' from
+-- "Data.List" the generating function @f@ cannot  dictate the length of the
+-- resulting vector, it must be statically known.
+--
+-- a simple use of 'unfoldrI':
+--
+-- >>> unfoldrI (\s -> (s,s-1)) 10 :: Vec 10 Int
+-- <10,9,8,7,6,5,4,3,2,1>
+unfoldrI :: KnownNat n => (s -> (a,s)) -> s -> Vec n a
+unfoldrI f s0 = map fst xs
+ where
+  xs = init (f s0 `Cons` ws)
+  ws = map (f . snd) (lazyV xs)
+{-# INLINE unfoldrI #-}
```

Comment:
Any reason for adding `INLINE` here? Is that something Clash specific?

Reply:
It's not a primitive, so the more inlining GHC can do, the better.